### PR TITLE
feat: opamp  implementation for nodejs agent

### PR DESCRIPTION
**Pull request by blumamir**

### CODE REVIEW

**Comment by edeNFed** [comment]
nit: maybe we should put the files in `/agents/nodejs` instead of `/agent-nodejs`?
We will probably have a similar OpAMP implementation for Java, NodeJS, Python, so having nested directories will be prettier.

File: agent-nodejs/package.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  "description": "Odigos distribution of OpenTelemetry for Node.js",
+  "main": "build/src/index.js",
+  "repository": "https://github.com/odigos-io/opentelemetry-node",
+  "author": "Odigos Authors",
+  "license": "Apache-2.0",
+  "scripts": {
+    "compile": "tsc -p ."
+  },
+  "dependencies": {
+    "@bufbuild/protobuf": "^1.10.0",
```
Comment (edeNFed) [inline agent-nodejs/package.json:13]: Update path in `dependabot.yaml`

File: opampserver/pkg/deviceid/k8sattributes.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		return "", false
+	}
+
+	annotations := obj.GetAnnotations()
+	if annotations == nil {
+		// no annotations, so service name is not specified by user. fallback to workload name
+		return "", false
+	}
+
+	overwrittenName, exists := annotations["odigos.io/reported-name"]
```
Comment (RonFed) [inline opampserver/pkg/deviceid/k8sattributes.go:40]: consts.OdigosReportedNameAnnotation

File: opampserver/pkg/deviceid/k8sattributes.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+func NewK8sPodInfoResolver(logger logr.Logger, kubeClient *kubernetes.Clientset) *K8sPodInfoResolver {
+	return &K8sPodInfoResolver{
+		logger:     logger,
+		kubeClient: kubeClient,
+	}
+}
+
+func (k *K8sPodInfoResolver) getServiceNameFromAnnotation(ctx context.Context, name string, kind string, namespace string) (string, bool) {
```
Comment (RonFed) [inline opampserver/pkg/deviceid/k8sattributes.go:27]: I think these functions should idealy be under k8sutils and not opampserver

File: opampserver/protobufs/anyvalue.pb.go
Comment (RonFed) [inline opampserver/protobufs/anyvalue.pb.go:1]: Are these generated with `protoc`, if so maybe worth adding in the README how to generate them,
Are we using the proto defined in opAmp, or are we having some custom types?

File: opampserver/pkg/server/server.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+		}
+	}()
+
+	go func() {
+		ticker := time.NewTicker(HeartbeatInterval)
+		defer ticker.Stop() // Clean up when done
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Info("Shutting down live connections timeout monitor")
```
Comment (RonFed) [inline opampserver/pkg/server/server.go:163]: Is there any cleanup we need to do here in exit?

File: opampserver/pkg/deviceid/kubelet.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+// this function will query the current state of the kubelet device plugins
+// and map each relevant device id to the container details which is using it.
+// givin a device id, we can know which pod and container it is allocated to and work our way
+// to complete the info for workload name and kind.
+//
+// unfortunately, I did not find a way to get just one device id from the kubelet, we need
+// to list and iterate over all the devices to get a current snapshot of allocations on this node
+// and then look for the ones we are interested in.
+func (c *KubeletClient) DeviceIdsToContainerDetails() (map[string]*ContainerDetails, error) {
```
Comment (RonFed) [inline opampserver/pkg/deviceid/kubelet.go:57]: This also does not look like it relates specifically to opamp server, maybe worth moving to k8sutils

File: agent-nodejs/src/opamp/client-http.ts [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  private httpClient: AxiosInstance;
+  private logger = diag.createComponentLogger({
+    namespace: "@odigos/opentelemetry-node/opamp",
+  });
+
+  // promise that we can resolve async later on, which the detect function can return
+  private resourcePromiseResolver?: (resourceAttributes: Attributes) => void;
+
+  constructor(config: OpAMPClientHttpConfig) {
+    this.config = config;
```
Comment (edeNFed) [inline agent-nodejs/src/opamp/client-http.ts:37]: Should we put a timeout on the HTTP requests?

File: agent-nodejs/src/opamp/client-http.ts [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    this.OpAMPInstanceUidString = uuidv7();
+    this.OpAMPInstanceUid = new TextEncoder().encode(
+      this.OpAMPInstanceUidString
+    );
+    this.httpClient = axios.create({
+      baseURL: `http://${this.config.opAMPServerHost}`,
+      headers: {
+        "Content-Type": " application/x-protobuf",
+        Authorization: `DeviceId ${config.instrumentationDeviceId}`,
```
Comment (edeNFed) [inline agent-nodejs/src/opamp/client-http.ts:46]: I think calling the header `Authorization` is a little misleading.
This gives a sense of secure connection but obviously, it is very easy to manipulate. Maybe something like `X-Odigos-DeviceId`

File: agent-nodejs/src/opamp/client-http.ts [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    // if we got here, it means we run out of retries and did not return from the loop
+    this.logger.error(
+      `Failed to get remote resource attributes from OpAMP server after retries, continuing without them`
+    );
+    this.resourcePromiseResolver?.({
+      [SEMRESATTRS_SERVICE_NAME]: this.config.instrumentationDeviceId,
+    });
+  }
+
+  private async sendHeartBeatToServer() {
```
Comment (edeNFed) [inline agent-nodejs/src/opamp/client-http.ts:163]: Do we need the heartbeat to keep the connection alive or is it part of OpAMP spec? Maybe we can drop it if we are not doing anything on the heartbeat

File: odiglet/Dockerfile [hunk 9]
```diff
@@ -3,10 +3,15 @@ WORKDIR /python-instrumentation
 ADD odiglet/agents/python/requirements.txt .
 RUN mkdir workspace && pip install --target workspace -r requirements.txt
 
-FROM node:16 AS nodejs-builder
+FROM node:18 AS nodejs-builder
+ARG ODIGOS_VERSION
```
[the remark this answers is not in the document]
Comment (blumamir) [reply odiglet/Dockerfile:7]: Currently. I think we should do it for all agents 👍

File: opampserver/pkg/server/server.go (hunk 4 above)
Comment (blumamir) [reply opampserver/pkg/server/server.go:163]: good catch, added `server.Shutdown`

File: agent-nodejs/src/opamp/client-http.ts (hunk 8 above)
Comment (blumamir) [reply agent-nodejs/src/opamp/client-http.ts:163]: OpAMP server uses it to figure out if the process (or the connection) is down

File: opampserver/protobufs/anyvalue.pb.go
Comment (blumamir) [reply opampserver/protobufs/anyvalue.pb.go:1]: The instructions can be found here: https://github.com/odigos-io/opamp-spec

File: opampserver/pkg/deviceid/kubelet.go (hunk 5 above)
Comment (blumamir) [reply opampserver/pkg/deviceid/kubelet.go:57]: Yes, I suggest doing it later on

**Review by edeNFed — APPROVED** [review]

**Pull request merged**